neo/SmartContract/InteropService.cs

```diff
@@ -143,8 +143,39 @@ private static bool Runtime_GetTrigger(ApplicationEngine engine)
 
         internal static bool CheckWitness(ApplicationEngine engine, UInt160 hash)
         {
-            var _hashes_for_verifying = engine.ScriptContainer.GetScriptHashesForVerifying(engine.Snapshot);
-            return _hashes_for_verifying.Contains(hash);
+            if (engine.ScriptContainer is Transaction tx)
+            {
+                CosignerUsage usage = tx.Attributes
+                    .Where(p => p.Usage == TransactionAttributeUsage.Cosigner)
+                    .Select(p => p.Data.AsSerializable<CosignerUsage>())
+                    .FirstOrDefault(p => p.Account.Equals(hash));
+                if (usage is null) return false;
+                if (usage.Scopes.HasFlag(WitnessScope.Global))
+                    return true;
+                if (usage.Scopes.HasFlag(WitnessScope.CalledByEntry))
+                {
+                    if (engine.CallingScriptHash == engine.EntryScriptHash)
+                        return true;
+                }
+                if (usage.Scopes.HasFlag(WitnessScope.CustomContracts))
+                {
+                    if (usage.AllowedContracts.Contains(engine.CurrentScriptHash))
+                        return true;
+                }
+                if (usage.Scopes.HasFlag(WitnessScope.CustomGroups))
+                {
+                    var contract = engine.Snapshot.Contracts[engine.CallingScriptHash];
+                    // check if current group is the required one
+                    if (contract.Manifest.Groups.Select(p => p.PubKey).Intersect(usage.AllowedGroups).Any())
+                        return true;
+                }
+                return false;
+            }
+
+            // only for non-Transaction types (Block, etc)
+
+            var hashes_for_verifying = engine.ScriptContainer.GetScriptHashesForVerifying(engine.Snapshot);
+            return hashes_for_verifying.Contains(hash);
```

Comment:
I think we can just disallow the calling to `CheckWitness` from non-Transaction types.

Reply:
Isnt it used to validate a block type ? I thought at some point an.engine would be loaded over a block for check... perhaps this wont happen.

Reply:
Better leave this possibility to the future Erik.